File: src/utils/anomalyDetection.mjs
import { IsolationForest } from 'isolation-forest';

// Função para detectar anomalias em movimentações de estoque
export function detectAnomalies(data) {
  // Selecionar características relevantes
  const features = data.map(item => [item.quantity, new Date(item.timestamp).getTime()]);
  
  // Criar e treinar o modelo Isolation Forest
  const model = new IsolationForest();
  model.fit(features);
  const scores = model.scores();
  
  // Marcar anomalias
  const anomalies = data.filter((_, index) => scores[index] < 0.5);
  return anomalies;
}

// Exemplo de uso
const data = [
  { quantity: 10, timestamp: '2023-10-01' },
  { quantity: 20, timestamp: '2023-10-02' },
  { quantity: 30, timestamp: '2023-10-03' },
  { quantity: 1000, timestamp: '2023-10-04' },
  { quantity: 50, timestamp: '2023-10-05' },
  { quantity: 60, timestamp: '2023-10-06' },
  { quantity: 70, timestamp: '2023-10-07' }
];

const anomalies = detectAnomalies(data);
console.log("Anomalias detectadas:", anomalies);

File: src/utils/anomaly_detection.py
import pandas as pd
from sklearn.ensemble import IsolationForest

# Função para detectar anomalias em movimentações de estoque
def detect_anomalies(data: pd.DataFrame) -> pd.DataFrame:
    # Selecionar características relevantes
    features = data[['quantity', 'timestamp']]
    
    # Normalizar dados
    features['timestamp'] = features['timestamp'].astype('int64')
    
    # Criar e treinar o modelo Isolation Forest
    model = IsolationForest(contamination=0.05, random_state=42)
    data['anomaly'] = model.fit_predict(features)
    
    # Marcar anomalias
    anomalies = data[data['anomaly'] == -1]
    return anomalies

# Exemplo de uso
if __name__ == "__main__":
    # Dados de exemplo
    data = pd.DataFrame({
        'quantity': [10, 20, 30, 1000, 50, 60, 70],
        'timestamp': pd.to_datetime([
            '2023-10-01', '2023-10-02', '2023-10-03',
            '2023-10-04', '2023-10-05', '2023-10-06', '2023-10-07'
        ])
    })
    
    anomalies = detect_anomalies(data)
    print("Anomalias detectadas:")
    print(anomalies)
